### PR TITLE
Add Child Register Page

### PR DESCRIPTION
**Pull request by prodhan29**
- Add Child Register component
- Test case for child-list

### CODE REVIEW

**Comment by sarfaraz0075** [comment]
@hmakandi and @p-netm, Please review this. We need this to be merged to proceed further.

File: clients/core/src/containers/Child/List/tests/index.test.tsx [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+describe('containers/child/List/ChildList', () => {
+    beforeEach(() => {
+        jest.resetAllMocks();
+    });
+
+    it('renders without crashing', () => {
+        const wrapper = mount(<ChildList />);
+        expect(wrapper.length).toBe(1);
+    });
+});
```
Comment (p-netm) [inline clients/core/src/containers/Child/List/tests/index.test.tsx]: @prodhan29 can we add some more tests here? I feel like. We have not fully tested this component.
Maybe we could add a test for the container as well.

Comment (prodhan29) [reply clients/core/src/containers/Child/List/tests/index.test.tsx]: Added

**Comment by prodhan29** [comment]
please suggest where to put the dropdown interfaces and default option.

File: clients/core/yarn.lock [hunk 2]
```diff
@@ -1,12116 +0,0 @@
-# THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
```
Comment (p-netm) [inline clients/core/yarn.lock]: why was this file deleted.

File: clients/core/src/containers/pages/Child/List/tests/index.test.tsx [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+            totalRecords: 0,
+            setTotalRecords: mock,
+        };
+        const wrapper = mount(
+            <Provider store={store}>
+                <ConnectedChildList {...props} />
+                );
+            </Provider>,
+        );
+        expect(toJson(wrapper)).toMatchSnapshot();
```
[the remark this answers is not in the document]
Comment (prodhan29) [reply clients/core/src/containers/pages/Child/List/tests/index.test.tsx]: @p-netm I followed your footsteps though. [reference](https://github.com/OpenSRP/opensrp-web/blob/7469ff10764947ca31684864937f8da1bacb6e47/clients/core/src/containers/Clients/List/tests/index.test.tsx#L62)

**Comment by prodhan29** [comment]
@p-netm  snapshot of the child-list page
image

File: clients/core/src/configs/env.ts [hunk 4]
```diff
@@ -112,3 +112,10 @@ export type OPENSRP_API_BASE_URL = typeof OPENSRP_API_BASE_URL;
 /** the clients endpoint NOTE: does not end with / */
 export const OPENSRP_CLIENT_ENDPOINT = process.env.REACT_APP_OPENSRP_CLIENT_ENDPOINT || 'client/search';
 export type OPENSRP_CLIENT_ENDPOINT = typeof OPENSRP_CLIENT_ENDPOINT;
+
+/** the clients endpoint NOTE: does not end with / */
+export const OPENSRP_CHILD_ENDPOINT = process.env.REACT_APP_OPENSRP_CLIENT_ENDPOINT || 'client/searchByCriteria';
```
[the remark this answers is not in the document]
Comment (prodhan29) [reply clients/core/src/configs/env.ts]: @p-netm  in the `client/core/src/constants` file there is a list of internal urls. do you want me to move all the enpoints to there ?

File: clients/core/src/containers/pages/Child/List/tests/index.test.tsx [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+            fetchChild: mock,
+            opensrpService: classMock,
+            removeChild: mock,
+            totalRecords: 0,
+            setTotalRecords: mock,
+        };
+        const wrapper = mount(<ChildList {...props} />);
+        expect(toJson(wrapper.find('Ripple'))).toMatchSnapshot('Ripple Loader');
+        wrapper.unmount();
+    });
```
[the remark this answers is not in the document]
Comment (prodhan29) [reply clients/core/src/containers/pages/Child/List/tests/index.test.tsx]: @p-netm  no. removed one

File: clients/core/src/containers/pages/Child/List/tests/index.test.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (prodhan29) [reply clients/core/src/containers/pages/Child/List/tests/index.test.tsx]: updated the test cases

File: clients/core/src/containers/pages/Child/List/index.tsx [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+/** props Interface for the clientList component */
+export interface ChildListProps {
+    opensrpService: typeof OpenSRPService;
+    childArray: Child[];
+    fetchChild: typeof fetchChildList;
+    removeChild: typeof removeChildList;
+    totalRecords: number;
+    setTotalRecords: typeof setTotalRecords;
+}
+
```
[the remark this answers is not in the document]
Comment (prodhan29) [reply clients/core/src/containers/pages/Child/List/index.tsx]: @p-netm done

File: clients/core/src/containers/pages/Child/List/index.tsx [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+    selectedGender: { value: '', label: 'All' },
+    loading: true,
+    searchText: '',
+};
+
+export interface ChildTableProps {
+    tableData: Child[];
+}
+
+function ChildTable(props: ChildTableProps): React.ReactElement {
```
[the remark this answers is not in the document]
Comment (prodhan29) [reply clients/core/src/containers/pages/Child/List/index.tsx]: @p-netm done

File: clients/core/src/containers/pages/Child/List/index.tsx [hunk 8]
```diff
@@ -0,0 +1,199 @@
...
+        });
+    };
+
+    genderFilter = (selectedGender: DropdownOption) => {
+        this.setState(
+            {
+                ...this.state,
+                selectedGender,
+            },
+            () => {
```
[the remark this answers is not in the document]
Comment (prodhan29) [reply clients/core/src/containers/pages/Child/List/index.tsx]: @redoy11  @p-netm, in this case, from where should i make the server call ? please suggest with an example.

File: clients/core/src/containers/pages/Child/List/index.tsx [hunk 9]
```diff
@@ -0,0 +1,199 @@
...
+                        </Col>
+                    </Row>
+                </div>
+            );
+        }
+    }
+}
+
+export { ChildList };
+/** Maybe define default props */
```
[the remark this answers is not in the document]
Comment (prodhan29) [reply clients/core/src/containers/pages/Child/List/index.tsx]: @p-netm Removed

File: clients/core/src/containers/pages/Child/List/helpers/tableDefinition.tsx [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+                        Header: 'Risk',
+                        accessor: '',
+                        disableSortBy: true,
+                    },
+                    {
+                        Header: 'Immunization status',
+                        accessor: '',
+                        disableSortBy: true,
+                    },
+                    // {
```
[the remark this answers is not in the document]
Comment (prodhan29) [reply clients/core/src/containers/pages/Child/List/helpers/tableDefinition.tsx]: @redoy11  @p-netm in the child table , each row has link to profile. how should i map that using the js configuration. please provide an example or update the code so that we can move faster to close this pr.

File: clients/core/yarn.lock (hunk 2 above)
Comment (prodhan29) [reply clients/core/yarn.lock]: recovered

File: clients/core/src/containers/pages/Child/List/index.tsx (hunk 8 above)
[the remark this answers is not in the document]
Comment (p-netm) [reply clients/core/src/containers/pages/Child/List/index.tsx]: I am not quite sure , but i think the blockQuoted text above suggests the call you are making in the componentDidupdate is enough

File: clients/core/src/configs/env.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (p-netm) [reply clients/core/src/configs/env.ts]: yes please.

File: clients/core/src/containers/pages/Child/List/helpers/tableDefinition.tsx (hunk 10 above)
[the remark this answers is not in the document]
Comment (p-netm) [reply clients/core/src/containers/pages/Child/List/helpers/tableDefinition.tsx]: Prerequisites: 
  - in the root folder that is the repo's root folder; run `yarn add -D -W @types/react-table`

then:

```typescript
{
                        // import { UseTableCellProps } from 'react-table';
                        // eslint-disable-next-line react/display-name
                        Cell: ({ row: { values } }: UseTableCellProps<Child>): ReturnType<React.FC> => (
                            <Link to={`${'CHILD_PROFILE_URL'}/${values.baseEntityId}`}>View</Link>
                        ),
                        Header: 'Actions',
                        accessor: '',
                        disableSortBy: true,
                    }
```

[the remark this answers is not in the document]
Comment (p-netm) [reply clients/core/src/containers/pages/Child/List/helpers/tableDefinition.tsx]: `CHILD_PROFILE_URL` is the url for the profile pages.

[the remark this answers is not in the document]
Comment (p-netm) [reply clients/core/src/containers/pages/Child/List/helpers/tableDefinition.tsx]: also view is a string that needs to go to constants

**Comment by p-netm** [comment]
@prodhan29 please fix the page header. it says `All clients`

File: clients/core/src/containers/pages/Child/List/index.tsx (hunk 8 above)
[the remark this answers is not in the document]
Comment (prodhan29) [reply clients/core/src/containers/pages/Child/List/index.tsx]: @redoy11  @p-netm , are you suggesting to write code like below ? if not then i am not getting your point. please provide a sample code in that case.

`componentDidUpdate() {
         this.getDataFromServer();
    };

    genderFilter = (selectedGender: DropdownOption) => {
        this.setState(
            {
                ...this.state,
                selectedGender,
            });
    };
    searchTextfilter = (searchText: string) => {
        this.setState(
            {
                ...this.state,
                searchText,
            });
    };`

[the remark this answers is not in the document]
Comment (p-netm) [reply clients/core/src/containers/pages/Child/List/index.tsx]: @prodhan29 yes.

**Comment by prodhan29** [comment]
> @prodhan29 please fix the page header. it says `All clients`

it is fixed.

File: clients/core/src/containers/pages/Child/List/index.tsx (hunk 8 above)
[the remark this answers is not in the document]
Comment (redoy11) [reply clients/core/src/containers/pages/Child/List/index.tsx]: @prodhan29  we do not need to use componentDidUpdate on this.  I think simply call the function this.getDataFromServer() after setState. Hope this is okay with you @p-netm.

[the remark this answers is not in the document]
Comment (redoy11) [reply clients/core/src/containers/pages/Child/List/index.tsx]: @p-netm, what will be the action on this? If we use componentDidUpdate(), what will be the breaking condition on this? The only thing that I can come up with is - we can use two sets of the same filter parameters - the request set and the reflected set. So if a change in gender filter occurs, it will setState on the request set. Then on componentDidUpdate(), we monitor the two set of filters for change and if change is present, we call the server for data with the values from the request set and also set the reflected set with the ones from the request set. Are you okay with this?

[the remark this answers is not in the document]
Comment (prodhan29) [reply clients/core/src/containers/pages/Child/List/index.tsx]: @p-netm we can use the `prevState` from one of the default parameters of the componentDidUpdate method to compare the filter value then we can go for the server call.

File: clients/core/src/containers/pages/Child/List/helpers/tableDefinition.tsx (hunk 10 above)
[the remark this answers is not in the document]
Comment (prodhan29) [reply clients/core/src/containers/pages/Child/List/helpers/tableDefinition.tsx]: @p-netm implemented

File: clients/core/src/containers/pages/Child/List/index.tsx (hunk 8 above)
[the remark this answers is not in the document]
Comment (p-netm) [reply clients/core/src/containers/pages/Child/List/index.tsx]: Ill put this down as a nitpicky kind of thing, since it is one way of achieving the same end.
